### DISCUSSION TITLE
Check class init arguments:  test with checkify?

### DISCUSSION THREAD

**Discussion by jecampagne** [jax-ml/jax#21943] [Q&A]
Hi, I have a long standing pb to perform a test on arguments of class __init__ for PyTrees.
Here a use case where Params has two arguments (a,b) and I would like to test that a>=0..

See below that checkify does not detect the second object instantiation error.

Have you an idea to do such test (nb. for the moment we leave to the user the responsability to instantiate with right args)

Thanks.

```python
import jax
jax.__version__
jax.config.update("jax_enable_x64", True)
from jax import random
import jax.numpy as jnp
import numpy as np

from jax.tree_util import register_pytree_node_class
from jax.experimental import checkify

@register_pytree_node_class
class Params:
    def __init__(self, a, b):
        self._a = a
        self._b = b
        jax.debug.print("__init__ Params(a={}, b={})",self._a,self._b)
        # the if code imply a TraceBack  
        #if a<0.0:
        #  raise ValueError("a must be >= 0")

        def check(a):
          checkify.check(a < 0.0, "a needs to be non-negative")
        jittable_check = checkify.checkify(check)
        jittable_check(a)

    def __repr__(self):
        return f"Params(a={self._a}, b={self._b})"

    @property
    def a(self):
        return self._a

    @property
    def b(self):
        return self._b

    def tree_flatten(self):
        children = (self._a, self._b)
        aux_data = None
        jax.debug.print("tree_flatten...")
        return (children, aux_data)

    @classmethod
    def tree_unflatten(cls, aux_data, children):
        jax.debug.print("tree_unflatten...")
        return cls(*children)
```
and I want to pass a jit test:
```python
jax.clear_caches()
identity = jax.jit(lambda x: x)
def test_jitting():
    objects = [
        Params(a=0.,b=1.1),
        Params(a=-1.0,b=1.1),
    ]

    # Test equality function
    def test_eq(self, other):
        return (
            self.a == other.a
            and self.b == other.b
        )

    # Check that after jitting the oject is still the same
    assert all([test_eq(identity(o), o) for o in objects])

test_jitting()
```

I get this result w/o any error detected:
```
__init__ Params(a=0.0, b=1.1)
__init__ Params(a=-1.0, b=1.1)
tree_flatten...
tree_flatten...
tree_flatten...
tree_unflatten...
__init__ Params(a=0.0, b=1.1)
tree_flatten...
tree_flatten...
tree_unflatten...
__init__ Params(a=0.0, b=1.1)
tree_unflatten...
__init__ Params(a=0.0, b=1.1)
tree_flatten...
tree_flatten...
tree_flatten...
tree_flatten...
tree_unflatten...
__init__ Params(a=-1.0, b=1.1)
tree_flatten...
tree_flatten...
tree_unflatten...
__init__ Params(a=-1.0, b=1.1)
tree_unflatten...
__init__ Params(a=-1.0, b=1.1)
tree_flatten...
```

**Comment by jecampagne** [comment]
A small up if someone is interested. Thanks

**Reply by patrick-kidger** [reply]
Try [`equinox.error_if`](https://docs.kidger.site/equinox/api/errors/) -- we use this to catch errors in `__init__` methods all the time.

**Reply by jakevdp** [reply]
If I understand correctly, this basically bottoms-out on `pure_callback`, yes? Do you see any performance impacts due to the process waiting for host-side communication and/or preventing fusions in the compiler?

**Reply by patrick-kidger** [reply]
So the callback is used solely to `raise` the error -- the predicate is computed JAX-side. The callback is wrapped in a `lax.cond`. So "no" to waiting on host-side communication, but "yes" to preventing fusion.

In the spirit of completeness I am aware of two other possible overheads here:
- `lax.cond` sometimes introduces spurious copies of its arguments. (In our case, of the token the error is threaded on to.)
- CSE stops at `jax.pure_callback`, even if it is two calls to the same function with the same argument.

**Reply by jakevdp** [reply]
But under `vmap`, `cond` will become `select` and so you'll have to wait for host-side communication regardless, yes?

**Reply by patrick-kidger** [reply]
Nope, I'm happy to say! We have an `unvmap_any` on the predicate:
https://github.com/patrick-kidger/equinox/blob/d3a01c8d1b5531f193f11514c0189104b8f62136/equinox/_errors.py#L272
That is, we raise an error if we are true for any batch element. The cond itself remains un-`select`ed.

(Most of the `lax.cond`s I write either have an unvmap before them for efficiency -- plus the appropriate logic to inside to behave correctly -- or an `eqxi.nonbatchable` to assert that we don't hit the `select` path.)

**Reply by jakevdp** [reply]
Do you do anything to prevent the compiler from DCE-ing your callbacks? `pure_callback` is explicitly assumed to be pure, meaning that if the output is unused in your program, the compiler can (and in general will) eliminate it. Perhaps `io_callback` or `debug.callback` would be a more robust mechanism for something like this?

**Reply by patrick-kidger** [reply]
In our case that's a feature, not a bug! I explicitly want these to be DCE-able if it turns out the later program doesn't need the result. (I even have a `check_dce` feature built on top of this, to assert that the compiler *has* DCE'd something!)

However it *is* a semi-common user-footgun to forget to use the return value, which I don't love, but haven't found a neat way to avoid. (Hacky but a good heuristic may just be to walk the stack, grab the source code, and take a look-see ourselves. Use that to generate a warning during trace time if necessary. Alternatively some kind of linear typing, if at some point JAX supports that for tracers?)

**Reply by jecampagne** [reply]
Ho Amazing!
I'm happy 
1) first because I see that I probably trigger some interesting discussions between you two (and I must say that I do not understand all that you've written),
 2) second it sounds that there is a solution :) I will then modify my code above to others see how to proceed.

**Reply by jakevdp** [reply]
This isn’t a great solution to be honest: it intentionally misuses `pure_callback` with a non-pure function (raising an error is a side-effect) and relies on implementation details that will not be guaranteed to continue working in future JAX versions. So if it works for you, great: but I’d suggest using it sparingly and not in the context of important code.

**Reply by patrick-kidger** [reply]
Yup, I'm aware! Unfortunately JAX doesn't have any other solution to eagerly raise errors!

**Reply by jakevdp** [reply]
True, this is because raising errors at runtime is not well-supported by XLA. You can hack around that limitation (as in `error_if`), but it's not an approach we should be recommending to users without large caveats.

**Reply by jecampagne** [reply]
@jakevdp Ok I note your remark. Do you know if JAX  will propose something to check arguments of PyTree is the context I have?

**Reply by jakevdp** [reply]
We have some ideas about that for the future, but fundamentally, no: there's currently no supported way to raise errors from XLA computations at runtime. It's why, for example, JAX handles out-of-bound indices by clipping rather than raising an error (e.g. [JAX Sharp Bits: out of bound indices](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#out-of-bounds-indexing)).

The best way forward is probably to embrace this quirk of JAX/XLA, and write your code accordingly.

**Reply by jecampagne** [reply]
Hum, "be silent" as the user provides wrong set of arguments is not what I was used to. Anyway, for the current version of our project we have clearly state to the future users the correct domain of validity of the functions.

Now, I was wandering if all the users (and library) already using Equinox (and probably error_if) are aware of the misuses of `pure_callback`, as en end user I guess no as it is rather tricky, no?

**Comment by jecampagne** [comment]
So here is the solution thanks to Equinox:
```python
import equinox as eqx

@register_pytree_node_class
class Params:
    def __init__(self, a, b):
        jax.debug.print("__init__ Params(a={}, b={})",a,b)

        a = eqx.error_if(a, a < 0, "a must be >= 0")
        self._a = a
        self._b = b 

    def __repr__(self):
        return f"Params(a={self._a}, b={self._b})"


    @property
    def a(self):
        return self._a

    @property
    def b(self):
        return self._b

    def tree_flatten(self):
        children = (self._a, self._b)
        aux_data = None
        jax.debug.print("tree_flatten...")
        return (children, aux_data)

    @classmethod
    def tree_unflatten(cls, aux_data, children):
        jax.debug.print("tree_unflatten...")
        return cls(*children)
```

Here is the prints and Traceback
```
_init__ Params(a=0.0, b=1.1)
__init__ Params(a=-1.0, b=1.1)

---------------------------------------------------------------------------

EquinoxTracetimeError                     Traceback (most recent call last)

[<ipython-input-21-aff0b188db24>](https://localhost:8080/#) in <cell line: 60>()
     58 
     59 
---> 60 test_jitting()

4 frames

[/usr/local/lib/python3.10/dist-packages/equinox/_errors.py](https://localhost:8080/#) in branched_error_if_impl(x, pred, index, msgs, on_error)
    283                     assert type(index) is int
    284                     if on_error == "raise":
--> 285                         raise EquinoxTracetimeError(msgs[index])
    286                     elif on_error == "breakpoint":
    287                         print(msgs[index])

EquinoxTracetimeError: a must be >= 0
```

**Reply by patrick-kidger** [reply]
I'm glad this has worked out for you!

And heads-up that you might also like `eqx.Module` as a more-robust alternative to registering your own pytree class. (Subclassing `eqx.Module` does this for you, whilst also avoiding several common mistakes -- e.g. your unflatten rule is calling `cls(...)`, which [won't work with vmap](https://jax.readthedocs.io/en/latest/working-with-pytrees.html#custom-pytrees-and-initialization-with-unexpected-values) etc)

**Reply by jecampagne** [reply]
Sounds interesting too :+ 
I do not know how to send u a private message to give you a link.

**Reply by jecampagne** [reply]
BTW:
if I had to plug such condition to trigger an error like that

```
if self._a == 0. and self._b <= 1:
  raise "for b<=1, then make a>0"
```
 
is there a way to do it with equinox error_if?

**Reply by patrick-kidger** [reply]
Something like
`self.a, self.b = eqx.error_if((a, b), (a == 0) && (b <= 1), "some message")`
I think!

**Reply by jecampagne** [reply]
Yes !!! it works :)
For instance
```python
eqx.error_if(None, a==0.0 and  b<=1.0, "Need a>0.0 for b<=1.0")
```
Sorry our msgs were made at the same times...
